packages/react-storage/src/components/StorageBrowser/Views/Controls/AddFiles.tsx

```diff
@@ -0,0 +1,96 @@
+import React from 'react';
+import { withBaseElementProps } from '@aws-amplify/ui-react-core/elements';
+
+import type { OmitElements } from '../types';
+import { StorageBrowserElements } from '../../context/elements';
+import { CLASS_BASE } from '../constants';
+import { useControl } from '../../context/controls';
+
+const { Button: ButtonElement } = StorageBrowserElements;
+
+const BLOCK_NAME = `${CLASS_BASE}__add-files`;
+
+export interface _AddFilesControl<
+  T extends StorageBrowserElements = StorageBrowserElements,
+> {
+  (): React.JSX.Element;
+  Button: T['Button'];
+}
+
+export interface AddFilesControl<
+  T extends StorageBrowserElements = StorageBrowserElements,
+> extends OmitElements<_AddFilesControl<T>, 'Button'> {
+  (): React.JSX.Element;
+}
+
+const Button = withBaseElementProps(ButtonElement, {
+  className: `${BLOCK_NAME}`,
+  variant: 'primary',
+});
+
+export const AddFilesControl: AddFilesControl = () => {
+  const [{ history }] = useControl({ type: 'NAVIGATE' });
+  const [{ selected }, handleUpdateState] = useControl({
+    type: 'ACTION_SELECT',
+  });
+
+  const { items: prevSelectedItems } = selected;
+
+  const fileUploadRef = React.useRef<HTMLInputElement>(null);
+
+  const destination = history[history.length - 1];
+
+  const handleInputChange = () => {
+    if (fileUploadRef.current?.files) {
+      const files: FileList = fileUploadRef.current?.files;
+      const items = prevSelectedItems ? [...prevSelectedItems] : [];
+      for (const data of files) {
+        const { name, lastModified, size } = data;
+
+        // if same file name is already in the list, replace it
+        if (prevSelectedItems?.find((item) => item.key === name)) {
+          const index = items.findIndex((item) => item.key === name);
+          items.splice(index, 1);
+        }
+
+        items.push({
+          key: name,
+          data,
+          lastModified: new Date(lastModified),
+          size,
+          type: 'FILE',
+        });
+      }
+
+      handleUpdateState({
+        actionType: 'UPLOAD_FILES',
+        type: 'SELECT_ACTION_TYPE',
+        destination,
+        items,
+        name: 'Upload Files',
+      });
```

Comment:
Not sure if this correct? Shouldn't be selecting an action if already in the action view

Reply:
yea, I wasn't sure either. I saw there is a type for `SET_UPLOADS_ITEMS` in the `ActionSelect` context. Would it be okay for me to update the reducer so that this type can update the `selected.items`?